metaflow_extensions/torchrun/plugins/torchrun_libs/executor.py

```diff
@@ -70,22 +70,46 @@ def _exec_cmd(
         self._ensure_torch_installed()
  
         # Container to build up the command to be run in a subprocess.
-        cmd = [sys.executable, "-m", "torch.distributed.run", f"--nproc_per_node={nproc_per_node}"]
-
-        # Construct the torchrun distributed arguments.
+        cmd = [sys.executable, "-m", "torch.distributed.run"]
 
-        # Group 1: args user supplied in torch.current.run
-        if type(torchrun_args) == dict:
-            self.torchrun_args.update(torchrun_args)
-            torchrun_args = _dict_to_args(self.torchrun_args)
-        cmd.extend(torchrun_args)
-
-        # Group 2: defaults user did not specify
-        for key, value in self.torchrun_args.items():
-            if f'--{key}' not in cmd:
-                cmd.extend([f'--{key}', str(value)])
-        cmd.extend(["--rdzv_endpoint", "%s:%s" % (self.torchrun_args["master_addr"], self.torchrun_args["master_port"])])
+        # Handle torchrun_args accounting for nproc_per_node
+        active_args = self.torchrun_args.copy()
+        active_args["nproc_per_node"] = nproc_per_node
 
+        # Construct the torchrun distributed arguments.
+        if isinstance(torchrun_args, dict):
+            active_args.update(torchrun_args)
+            cmd_args = _dict_to_args(active_args)
+            cmd.extend(cmd_args)
+        else:
+            # If torchrun_args is a list, we need to check for existing args and supplement with defaults
+            processed_args = list(torchrun_args)  # Create a copy
+            override_args = set()
+            
+            # First, check which args are already in the list and update nproc_per_node if present
+            for i, arg in enumerate(processed_args):
+                if arg.startswith('--'): 
+                    arg_name = arg[2:]
+                    override_args.add(arg_name)
+                    if arg_name == 'nproc_per_node' and i+1 < len(processed_args):
+                        processed_args[i+1] = str(nproc_per_node)
+            
+            # Then add any default args not already in the list
+            for arg_key, arg_val in active_args.items():
+                if arg_key not in override_args:
+                    processed_args.extend([f'--{arg_key}', str(arg_val)])
+            
```

Comment:
Sets are not ordered so you might change arg order if you do set conversions. Why not do the following:

Check for the nproc-per-node in the list. If it is not present then append it otherwise do nothing ?

Reply:
You don't need to check for the subsequent value of nproc-per-node, just that cli arg is present